component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java

```diff
@@ -0,0 +1,388 @@
+            }
+            maxHeightOfTree -= 1; // since this parameter when set to 0 MOA ClusTree implementation builds one root
+            // and 3 children. i.e one level
+
+            //expressionExecutors[4] --> horizon
+            if (!(attributeExpressionExecutors[4] instanceof ConstantExpressionExecutor)) {
+                throw new SiddhiAppCreationException("horizon has to be a constant but found " +
+                        this.attributeExpressionExecutors[4].getClass().getCanonicalName());
+            }
+
+            if (attributeExpressionExecutors[4].getReturnType() == Attribute.Type.INT) {
+                horizon = (Integer) ((ConstantExpressionExecutor)
+                        attributeExpressionExecutors[4]).getValue();
+                if (horizon <= 0) {
+                    throw new SiddhiAppCreationException("horizon should be a positive integer " +
+                            "but found " + horizon);
+                }
+            } else {
+                throw new SiddhiAppCreationException("horizon should be of type int but found " +
+                        attributeExpressionExecutors[4].getReturnType());
+            }
+        }
+
+        noOfDimensions = attributeExpressionExecutors.length - attributeStartIndex;
+        coordinateValuesOfCurrentDataPoint = new double[noOfDimensions];
+
+        //validating all the attributes to be variables
+        featureVariableExpressionExecutors = CoreUtils.extractAndValidateFeatures(inputDefinition,
+                attributeExpressionExecutors, attributeStartIndex, noOfDimensions);
+
+        //creating models
+        clusTreeModel = new ClusTreeModel();
+        clusTreeModel.init(noOfDimensions, noOfClusters, maxHeightOfTree, horizon);
+        kMeansModel = new KMeansModel();
+
+        executorService = siddhiAppContext.getExecutorService();
+
+        //setting return attributes
+        List<Attribute> attributeList = new ArrayList<>(1 + noOfDimensions);
+        attributeList.add(new Attribute("euclideanDistanceToClosestCentroid", Attribute.Type.DOUBLE));
+        for (int i = 1; i <= noOfDimensions; i++) {
+            attributeList.add(new Attribute("closestCentroidCoordinate" + i, Attribute.Type.DOUBLE));
+        }
+        return attributeList;
+    }
+
+    @Override
+    protected void process(ComplexEventChunk<StreamEvent> complexEventChunk, Processor processor,
+                           StreamEventCloner streamEventCloner, ComplexEventPopulater complexEventPopulater) {
+        synchronized (this) {
+            while (complexEventChunk.hasNext()) {
+                StreamEvent streamEvent = complexEventChunk.next();
+                noOfEventsReceived++;
+
+                //validating and getting coordinate values
+                for (int i = 0; i < noOfDimensions; i++) {
+                    try {
+                        Number content = (Number) featureVariableExpressionExecutors.get(i).execute(streamEvent);
```

Comment:
Can't we do this validation at init() by using getReturnType() method in ExpresionExecutor?

Reply:
we are doing the validation of the query feature type in init. but this is done to check if the user fails to honor the stream definition and pass a non-numeric into the inputhandler